src/types/params.ts

```diff
@@ -2,15 +2,15 @@ export type ParamExtras = {
   invalid: (message?: string) => never,
 }
 
-export type ParamGetter<T = any> = (value: string, extras: ParamExtras) => T
-export type ParamSetter<T = any> = (value: T, extras: ParamExtras) => string
+export type ParamGetter<T = any> = (value: string | undefined, extras: ParamExtras) => T
+export type ParamSetter<T = any> = (value: T, extras: ParamExtras) => string | undefined
```

Comment:
Seeing this now I'm second guessing adding the `undefined` to ParamGetter. A param getter should never be called with an undefined value right?

I also think the same for ParamSetter. It should always return a string. 

Optional values shouldn't ever pass through userland get/set methods. 

These are purely to support the `optional` utility yeah? Wonder if there's a clean way to support that without adding `undefined` to the exported types.

Reply:
no worries. I think it will work out just as well to use empty strings as being synonymous with `undefined`